### ISSUE TITLE
Invocation to destroy on a TransactionalMap in a client affects directly to the cluster

### ISSUE THREAD

**Issue by alexrg91** [hazelcast/hazelcast#16835]
**Describe the bug**
Maps destroyed in a client transaction context are also directly destroyed in the server.

I'm trying to clear a map during a transaction prior to refill it. We are migrating from Hazelcast 2.6 to 3.12 and adapting when necessary. In the previous version they used map.destroy() to clear a map inside a transaction and it doesn't cause any problem. Maps on the cluster keep existing until the commit is done and then the maps are updated.
As far as I've tryed, when I try to use the destroy method for a TransactionalMap it also destroys the map on the cluster.

I have searched over all the documentation I have found but I haven't seen any reference where it say destroy can not be used inside a transaction. 
Is this intended to work this way?

**Expected behavior**
If destroy is invoked on a TransactionalMap during a transaction the map on the cluster won't be affected until commit is done.

**To Reproduce**
I have done a simple JUnit to test the problem.

	

	import static org.junit.Assert.assertTrue;

	import org.junit.After;
	import org.junit.Before;
	import org.junit.Test;

	import com.hazelcast.client.HazelcastClient;
	import com.hazelcast.core.Hazelcast;
	import com.hazelcast.core.HazelcastInstance;
	import com.hazelcast.core.IMap;
	import com.hazelcast.core.TransactionalMap;
	import com.hazelcast.transaction.TransactionContext;
	import com.hazelcast.transaction.TransactionOptions;

	public class DemoHzcApplicationTests {
		
		private static final String MAP_TEST = "prueba";
		
		private HazelcastInstance server;
		private HazelcastInstance client;
		
		private TransactionOptions options = new TransactionOptions().setTransactionType(TransactionOptions.TransactionType.TWO_PHASE);
		
		@Before
		public void init() {
			server = Hazelcast.newHazelcastInstance();
			client = HazelcastClient.newHazelcastClient();
		}

		@Test
		public void test1() {
			final String key1 = "key1";
			final String value1 = "value1";
			final String key2 = "key2";
			final String value2 = "value2";
			final String key3 = "key3";
			final String value3 = "value3";
			IMap<Object, Object> mapTest = this.client.getMap(MAP_TEST);
			mapTest.put(key1, value1);
			
			this.client.getMap(MAP_TEST).keySet().forEach(System.out::println);
			assertTrue(this.client.getMap(MAP_TEST).get(key1).equals(value1));
			
			TransactionContext newTransactionContext = this.client.newTransactionContext(this.options);
			newTransactionContext.beginTransaction();
			TransactionalMap<Object, Object> transactionalMap = newTransactionContext.getMap(MAP_TEST);
			transactionalMap.put(key2, value2);
			// The map on the server should have just 1 entry because transaction has not been committed
			assertTrue(this.server.getMap(MAP_TEST).size() == 1);
			newTransactionContext.commitTransaction();
			// The map on the server should have 2 entries after the commit
			assertTrue(this.client.getMap(MAP_TEST).size() == 2);
			
			
			TransactionContext newTransactionContext2 = this.client.newTransactionContext(this.options);
			newTransactionContext2.beginTransaction();
			TransactionalMap<Object, Object> transactionalMap2 = newTransactionContext2.getMap(MAP_TEST);
			// Invocation to destroy
			transactionalMap2.destroy();
			transactionalMap2.put(key3, value3);
			// Since transaction has not been committed there should be the 2 previous keys but the map is empty
			assertTrue(this.server.getMap(MAP_TEST).size() == 2);
			newTransactionContext2.commitTransaction();
			
			assertTrue(this.client.getMap(MAP_TEST).get(key3).equals(value3));
			assertTrue(this.client.getMap(MAP_TEST).size() == 1);
		}

		@After
		public void after() {
			this.client.shutdown();
			this.server.shutdown();
		}
	}

Hazelcast version is 3.12.6
JDK 8

**Comment by mmedenjak** [comment]
Hi @alexrg91 ! I haven't tried the code myself but calling `destroy` in a transactional context seems wrong.

You say that that you are `trying to clear a map`. So calling `clear()`  seems better but unfortunately we don't support it yet for transactional map. You're free to submit a PR or we will keep this issue open and consider implementing it in the future.

**Comment by alexrg91** [comment]
Hi @mmedenjak ! Thanks for the quick reply.
My first intention was to use clear but, as you mention, it is not on the TransactionalMap Interface.
Since calling to destroy method affects directly to the distributed Object in the cluster, wouldn't be a possibility to throw a warning log or an exception?

I'll try to open a PR!

**Comment by mmedenjak** [comment]
Yes, we could log a warning or throw an exception but right now we can't modify the behaviour because we have to keep backwards compatibility. Some users might rely on the current behaviour.
